Allow JIT compilation for system emulation

vacantron opened the pull request:
This patch makes `rv32emu` support the just-in-time (JIT) compilation in system simulation. To achieve this goal, a "satp (supervisor address translation and protection)" field has been introduced to the block structure in JIT mode to ensure the block cache is replaced correctly.

The file `src/system.c` has been updated for the easier reuse of functions, and the MOP fusion and T2C are disabled temporarily. To boot the Linux Kernel, try the following commands:

```shell
$ make ENABLE_SYSTEM=1 ENABLE_MOP_FUSION=0 ENABLE_JIT=1 ENABLE_T2C=0
$ ./build/rv32emu -k <image> -i <rootfs> -b <dtb>
``` 
 <div id='description'>
<h3>Summary by Bito</h3>
This PR implements JIT compilation support for RISC-V system emulation by adding SATP tracking and MMU-aware instruction handling. The changes include refactoring MMIO operations into separate helper functions, implementing comprehensive instruction handlers with proper memory access handling, and enhancing the block caching system with improved error handling for Linux kernel boot compatibility.
<br>
<br>
<b>Unit tests added</b>: False
<br>
<br>
<b>Estimated effort to review (1-5, lower is better)</b>: 5
</div>

jserv reviewed (commented):
<!-- github-benchmark-action-comment(start): Benchmarks Summary -->
# Benchmarks

<details>

| Benchmark suite | Current: 4ef61b8966268a765455d5e30c5e0bdb18eca93e | Previous: 7a032dbb90098acdd4e602e62d7ead60b26fdb3b | Ratio |
|-|-|-|-|
| `Dhrystone` | `1284` Average DMIPS over 10 runs | `1340` Average DMIPS over 10 runs | `1.04` |
| `Coremark` | `972.508` Average iterations/sec over 10 runs | `947.06` Average iterations/sec over 10 runs | `0.97` |

</details>

This comment was automatically generated by [workflow](https://github.com/sysprog21/rv32emu/actions?query=workflow%3ABenchmark) using [github-action-benchmark](https://github.com/marketplace/actions/continuous-benchmark).
<!-- github-benchmark-action-comment(end): Benchmarks Summary -->

jserv commented on .gitignore:
```diff
@@ -28,4 +28,3 @@ tests/arch-test-target/config.ini
 tests/arch-test-target/sail_cSim/riscv_sim_RV32
 tests/scimark2/
 __pycache__/
-src/rv32_jit.c
```
Can you explain why `src/rv32_jit.c` is no longer generated?

vacantron replied:
> Can you explain why `src/rv32_jit.c` is no longer generated?

The commit message has been updated. Please check [30f01db](https://github.com/sysprog21/rv32emu/pull/521/commits/30f01db32cf02b4e657e1a6c56eea39d593a6977)

jserv replied:
> The commit message has been updated. Please check [30f01db](https://github.com/sysprog21/rv32emu/pull/521/commits/30f01db32cf02b4e657e1a6c56eea39d593a6977)

Considered as preparation of #511

jserv commented on src/emulate.c:
```diff
@@ -838,12 +841,11 @@ static block_t *block_find_or_translate(riscv_t *rv)
+#if RV32_HAS(SYSTEM)
+    if (next_blk && next_blk->satp != rv->csr_satp)
+        next_blk = NULL;
+#endif
```
Add comments for SATP.

jserv commented on src/emulate.c:
```diff
@@ -910,6 +918,32 @@ static block_t *block_find_or_translate(riscv_t *rv)
+         * However, if the structure "branch_table_t" is going to reference the
+         * block directly, this updating is nacessary to avoid to use the freed
+         * blocks.
+         */
```
I didn't get it. When is it safe to release the resources?

vacantron replied:
As the comments above, the current workflow is safe. However, if the implementation of `branch_table` is changed (e.g. storing the possible blocks directly instead of the program counters), this updating cannot be ignored.

jserv commented on src/riscv_private.h:
```diff
@@ -126,6 +129,18 @@ struct riscv_internal {
 
+#if RV32_HAS(JIT) && RV32_HAS(SYSTEM)
+    /*
+     * Aarch encoder only accepts 9 bits signed offset. Do not put this
```
Aarch?

jserv commented on src/riscv_private.h:
```diff
@@ -126,6 +129,18 @@ struct riscv_internal {
+     * structure to the bottom.
+     */
+    struct {
+        uint32_t is_mmio; /* whether is MMIO or not */
+        uint32_t type;    /* 0: read, 1: write */
```
Would you use bit-fields instead?

vacantron replied:
To simplify the emission of load/store instructions of both `x86-64` and `Aarch64` in `src/rv32_jit.c`, I would like to keep them as unsigned integer.

ChinYikMing commented on src/emulate.c:
```diff
@@ -704,6 +706,7 @@ static inline void remove_next_nth_ir(const riscv_t *rv,
  * Strategies are being devised to increase the number of instructions that
  * match the pattern, including possible instruction reordering.
  */
+#if RV32_HAS(MOP_FUSION)
```
Wrap around the comment as well.

ChinYikMing commented:
### Testing Report:

**x86-64 machine:**
Issue: kernel panic.

Kernel message Logs:
```
[    0.000000] Linux version 6.1.119 (mirv32@node11) (riscv32-buildroot-linux-gnu-gcc.br_real (Buildroot 2024.08.3) 14.2.0, GNU ld (GNU Binutils) 2.42) #1 Sat Dec 14 22:54:23 CST 2024
[    0.00000] Machine model: rv32emu
[   .00000] earlycon: ns16550 at MMIO 0xf4000000 (options '')
[   .00000] printk: bootcons1655] enabled
[    0.000000] Zone ranges:
[    0.00000]   Normal   [mem 0x0000000000000000-0x000000001fffffff]
[   .00000] Movable zone start for each node
[   .00000] Early memory node ranges
[   .00000]   node   0: [mem 0x000000000000000-0x000000001ffffff]
[   .00000] Initmem setup node 0 [mem 0x000000000000000-0x000000001ffffff]
[   .00000] SBI specification v0.3 detected
[   .00000] SBI implementation 99 Version
[   .00000] SBI TIME extension detected
[   .00000] SBI SRST extension detected
[    0.000000] riscv: base ISA extensions aim
[    0.00000] riscv: ELF capabilities aim
[   .00000] Bui zonelists, mobility grouping on.  Total13004
[   .00000] Kernel command line: earlycon console=ttyS0
[   .000000] Dentry cache hash table entries: 65536 (order: 6, 26214
[    0.00000] Inode-cache hash table en3276 (or13107 bytes, linear)
[   .00000] mem auto-init: stack:off, heap alloc:off, heap free:off
[   .00000] 2428K avail323K kernel c33K rwd83K rod16K i131742K reserveK cma-reserved)
[    0.000000] SLUB: HWalign=64, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[   .000000] NR_IRQS: 64, nr_irqs: 64, preallocated irqs: 0
[    0.00000] riscv-in3 local interrupts mapped
[    0.000000] plic: interrupt-controller@0: mapped 31 interrupts with 1 handlers for 1 contexts.
[    0.000000] riscv-timer: riscv_timer_init_dt: Registering clocksource cpuid [0] hartid [0]
[    0.00000] clocksource: riscv_clocksource: mask: 0xffffffffffffffff maxefdb196d, max_idle_ns: 440795204367 ns
[   .00000] sched_clo6 bits6MHz, resolut1n219902325555ns
[   .00068] Console: colour dummy device 80x25
[    0.00093] Calibrating delay loop (skipped), value calculated using timer frequency..
[   .00097] 00 BogoMI26000)
[   .00119] pid_max: de3276 mini30
[   .00203] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.00224] Mountpoint-cache hash table ent102 (orde409 bytes, linear)
[    0.00638] ASID allocator using 9 bits (512 entries)
[    0.008134] devtmpfs: initialized
[   .01275] clocksource: jiffies: mask: 0xffffffff max_fffffff, max_idle_ns: 7645041785100000 ns
[    0.013002] futex hash table entr25 (order: -1, 3072 bytes, l
[    0.015155] NET: Registered PF_NETLINK/PF_ROUTE protocol family
[    0.018778] platform soc@F0000000: Fixed dependency cycle(s) with /soc@F0000000/interrupt-controller@0
[    0.038375] clocksource: Switched to clocksource riscv_clocksource
[   .05459] sysfs: cannot create duplicate filename '/dev/char/4:'
[   .054745] CPU: 0 PID: 1 Comm: swapper Not tainted 6.1.119 #1
[    0.054941] Hardware name: rv32emu (DT)
[   .05504] Call Trace:
[   .05512] [<c0003ad>] dump_backtrace+03
[   .05531] [<c03228d>] show05
[   .05543] [<c032576>] dump_sta04
[   .05556] [<c032579>] dump02
[   .05568] [<c010e43>] sysfs_wa09
[   .05580] [<c010e76>] sysfs_do_create_l0c
[   .05594] [<c010e7a>] sysfs_creat04
[   .05606] [<c01fd13>] dev067
[   .05619] [<c01fd36>] device_re03
[   .05631] [<c01d117>] tty_register_devi021
[   .05645] [<c01d123>] tty_register_02
[   .05658] [<c01d13b>] tty_register01d
[   .05671] [<c033d8b>] v018
[   .05683] [<c033cd4>] t016
[   .05695] [<c033e6c>] chr_d010
[   .05708] [<c000110>] do_one_i026
[   .05720] [<c032c06>] kernel_init_f021
[   .05733] [<c0325a3>] kern011
[   .05746] [<c00023d>] ret_from_exc01
[    0.062338] Kernel panic - not syncing: Couldn't register console driver
[   .062494] CPU: 0 PID: 1 Comm: swapper Not tainted 6.1.119 #1
[    0.062688] Hardware name: rv32emu (DT)
[   .06279] Call Trace:
[   .06287] [<c0003ad>] dump_backtrace+03
[   .06306] [<c03228d>] show05
[   .06318] [<c032576>] dump_sta04
[   .06331] [<c032579>] dump02
[   .06343] [<c0322b8>] f
[   .06355] [<c033d7b>] v018
[   .06367] [<c033cd4>] t016
[   .06379] [<c033e6c>] chr_d010
[   .06391] [<c000110>] do_one_i026
[   .06404] [<c032c06>] kernel_init_f021
[   .06417] [<c0325a3>] kern011
[   .06429] [<c00023d>] ret_from_exc01
[   .06445] ---[ end Kernel panic - not syncing: Couldn't register console driver ]---
```

**Aarch64 machine:**
Issue: `Coremark` seems running forever, perhaps into a infinite loop. I have run about 30 mins but still no any output.

jserv reviewed (changes requested):
Rebase the latest master branch to resolve Arm64 host breakage on CI.

jserv reviewed (changes requested):
Rebase the latest master branch and resolve the conflicts.

vacantron committed:
Add "src/rv32_jit.c" into repository

The file is generated by "tools/gen-jit-template.py"

To bring up the Linux Kernel to just-in-time (JIT) compilation, we need
to update the memory-related operation with the memory management unit.
However, the current "src/rv32_jit.c" was generated by the template.
That template reduced the rework for the repeated statements, but also
reduced the flexibility and the intuitiveness for bring up the new
feature. In this commit, we deprecate that template and just use a
regular file for the flexibility.

vacantron committed:
Refactor "src/system.c" to enhence reusability

vacantron committed:
Fix missing LLVM libraries

jserv commented:
CI breakage: [Arm64 host](https://github.com/sysprog21/rv32emu/actions/runs/13223638322/job/36915001332)
```
src/jit.c: In function 'emit_jmp':
src/jit.c:1117:38: error: unused parameter 'target_satp' [-Werror=unused-parameter]
 1117 |                             uint32_t target_satp)
      |                             ~~~~~~~~~^~~~~~~~~~~
```

jserv commented:
CI breakge: [x86 host](https://github.com/sysprog21/rv32emu/actions/runs/13223638322/job/36915001135)
```
src/system.c: In function ‘mmu_read_w’:
src/system.c:221:53: error: ‘R’ undeclared (first use in this function)
  221 |     uint32_t addr = rv->io.mem_translate(rv, vaddr, R);
      |                                                     ^
src/system.c:221:53: note: each undeclared identifier is reported only once for each function it appears in
src/system.c: In function ‘mmu_read_s’:
src/system.c:240:53: error: ‘R’ undeclared (first use in this function)
  240 |     uint32_t addr = rv->io.mem_translate(rv, vaddr, R);
      |                                                     ^
src/system.c: In function ‘mmu_read_b’:
src/system.c:252:53: error: ‘R’ undeclared (first use in this function)
  252 |     uint32_t addr = rv->io.mem_translate(rv, vaddr, R);
      |                                                     ^
src/system.c: In function ‘mmu_write_w’:
src/system.c:271:53: error: ‘W’ undeclared (first use in this function)
  271 |     uint32_t addr = rv->io.mem_translate(rv, vaddr, W);
      |                                                     ^
src/system.c: In function ‘mmu_write_s’:
src/system.c:290:53: error: ‘W’ undeclared (first use in this function)
  290 |     uint32_t addr = rv->io.mem_translate(rv, vaddr, W);
      |                                                     ^
src/system.c: In function ‘mmu_write_b’:
src/system.c:305:53: error: ‘W’ undeclared (first use in this function)
  305 |     uint32_t addr = rv->io.mem_translate(rv, vaddr, W);
      |                                                     ^
```

jserv commented on src/utils.c:
```diff
@@ -204,10 +220,19 @@ bool set_add(set_t *set, uint32_t key)
  * @key: the key of the inserted entry
  */
+#if RV32_HAS(SYSTEM) && RV32_HAS(JIT)
+bool set_has(set_t *set, uint64_t key)
```
You should explain why wider key length is required.

vacantron replied:
It is at https://github.com/sysprog21/rv32emu/blob/b982d62a4c77ee44476f2f47de9aabbbf18ba27a/src/utils.h#L144-L150

Should I mention it again here?

jserv replied:
> Should I mention it again here?

Consider to utilize typedef and helper routines to unify.

jserv commented on src/utils.h:
```diff
@@ -13,16 +13,32 @@ void rv_gettimeofday(struct timeval *tv);
 
+#if RV32_HAS(JIT) && RV32_HAS(SYSTEM)
+
+typedef uint64_t rv_hash_dtype_t;
```
How about `rv_hash_t`? The suffix `_t` implies the type.

vacantron replied:
> How about `rv_hash_t`? The suffix `_t` implies the type.

I concern about it might be confused with the hash function or other data structure. It refers to the length of key, and the suffix `_t` denotes it is a alias defined by keyword `typedef` .

jserv replied:
> It refers to the length of key, and the suffix `_t` denotes it is a alias defined by keyword `typedef` .

Alternatively, use `rv_key_t`?

vacantron replied:
Sure. I prefer to use `rv_hash_key_t` to prevent name conflict.

vacantron committed:
Allow JIT compilation for system simulation

This commit introduces "satp" field to the block structure in JIT mode
to ensure the block cache is replaced correctly.

The MOP fusion and T2C are disabled temporarily. Use the following
commands to boot the Linux Kernel:

$ make ENABLE_SYSTEM=1 ENABLE_MOP_FUSION=0 ENABLE_JIT=1 ENABLE_T2C=0
$ ./build/rv32emu -k <image> -i <rootfs> -b <dtb>

jserv reviewed (changes requested):
Rebase the latest `master` branch and resolve the conflicts.

ChinYikMing commented:
Run JIT enabled system emulation verification in CI?

vacantron committed:
CI: Add JIT system simulation test

vacantron commented:
> Run JIT enabled system emulation verification in CI?

It had been added in [8d46115](https://github.com/sysprog21/rv32emu/pull/521/commits/8d461158145ccdac1f08a29f2b971bc4dc51eee2) .

jserv commented:
Thank @vacantron for contributing!
